localboot/grub.go

```diff
@@ -95,35 +126,64 @@ func ParseGrubCfg(grubcfg string, basedir string, grubVersion int) []bootconfig.
 	return bootconfigs
 }
 
+func unquote(ver grubVersion, text string) string {
+	if ver == grubV2 {
+		// if grub2, unquote the string, as directives could be quoted
+		// https://www.gnu.org/software/grub/manual/grub/grub.html#Quoting
+		// TODO unquote everything, not just \$
+		return strings.Replace(text, `\$`, "$", -1)
+	}
+	// otherwise return the unmodified string
+	return text
+}
+
+func isMn(r rune) bool {
+	return unicode.Is(unicode.Mn, r) // Mn: nonspacing marks
+}
+
 // ScanGrubConfigs looks for grub2 and grub legacy config files in the known
 // locations and returns a list of boot configurations.
 func ScanGrubConfigs(basedir string) []bootconfig.BootConfig {
 	bootconfigs := make([]bootconfig.BootConfig, 0)
-	// Scan Grub 2 configurations
-	for _, grubpath := range Grub2Paths {
-		path := path.Join(basedir, grubpath)
-		log.Printf("Trying to read %s", path)
-		grubcfg, err := ioutil.ReadFile(path)
+	err := filepath.Walk(basedir, func(currentPath string, info os.FileInfo, err error) error {
 		if err != nil {
-			log.Printf("cannot open %s: %v", path, err)
-			continue
+			return err
 		}
-		crypto.TryMeasureData(crypto.ConfigData, grubcfg, path)
-		cfgs := ParseGrubCfg(string(grubcfg), basedir, 2)
-		bootconfigs = append(bootconfigs, cfgs...)
-	}
-	// Scan Grub Legacy configurations
-	for _, grubpath := range GrubLegacyPaths {
-		path := path.Join(basedir, grubpath)
-		log.Printf("Trying to read %s", path)
-		grubcfg, err := ioutil.ReadFile(path)
-		if err != nil {
-			log.Printf("cannot open %s: %v", path, err)
-			continue
+		t := transform.Chain(norm.NFD, transform.RemoveFunc(isMn), norm.NFC)
+		currentPath, _, _ = transform.String(t, currentPath)
+		if path.Dir(currentPath) == basedir && info.IsDir() && !isGrubSearchDir(path.Base(currentPath)) {
+			debug("Skip %s", currentPath)
+			// skip irrelevant toplevel directories
+			return filepath.SkipDir
 		}
-		crypto.TryMeasureData(crypto.ConfigData, grubcfg, path)
-		cfgs := ParseGrubCfg(string(grubcfg), basedir, 1)
-		bootconfigs = append(bootconfigs, cfgs...)
+		if info.IsDir() {
+			debug("Check %s", currentPath)
+			// continue
+			return nil
+		}
+		cfgname := info.Name()
+		if cfgname == "grub.cfg" || cfgname == "grub2.cfg" {
```

Comment:
this block could be rewritten to be more clear. We also want an error if the grub file is found but cannot be read.
```
switch cfgname {
case "grub.cfg":
	ver = grubV1
case "grub2.cfg":
	ver = grubV2
default:
	return nil
}
log.Printf("Parsing %s", currentPath)
data, err := ioutil.ReadFile(currentPath)
if err != nil {
	return err
}
cfgs := ParseGrubCfg(ver, string(grubcfg), basedir)
bootconfigs = append(bootconfigs, cfgs...)
```

Reply:
This looks way more elegant ;) - changed it